### PR TITLE
[client] Multiple client one driver

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

This PR supports multi-client one driver mode.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
The error message is not that good for now, but I think it should be a good start.
```
>>> cli1 = ray.init("ray://localhost:10001", allow_multiple=True)
>>> cli2 = ray.init("ray://localhost:10001", allow_multiple=True)
>>> with cli1:
...     a = ray.put(10)
...
>>> with cli2:
...     a = ray.put(20)
...
>>> with cli1:
...     ray.get(a)
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/mnt/yic/ray-w2/python/ray/_private/client_mode_hook.py", line 81, in wrapper
    return getattr(ray, func.__name__)(*args, **kwargs)
  File "/mnt/yic/ray-w2/python/ray/util/client/api.py", line 42, in get
    return self.worker.get(vals, timeout=timeout)
  File "/mnt/yic/ray-w2/python/ray/util/client/worker.py", line 225, in get
    res = self._get(obj_ref, op_timeout)
  File "/mnt/yic/ray-w2/python/ray/util/client/worker.py", line 244, in _get
    err = cloudpickle.loads(data.error)
EOFError: Ran out of input
```

File: python/ray/client_builder.py
Comment (ericl) [inline python/ray/client_builder.py:112]: What example code reproduces this error? Seems confusing to raise an error if allow_multiple is True.

File: python/ray/client_builder.py
Comment (ericl) [inline python/ray/client_builder.py:117]:
```suggestion
            raise RuntimeError("Multiple clients are connected. Additional connections must be created with `ray.init(allow_multiple=True)`")
```

File: python/ray/util/client/__init__.py
Comment (ijrsvt) [inline python/ray/util/client/__init__.py:15]: nit: Can you name this `_ClientContext` or something?

File: python/ray/worker.py
Comment (ijrsvt) [reply python/ray/worker.py:578]: It is possible that tests are still using this hook.

File: python/ray/client_builder.py
Comment (ijrsvt) [reply python/ray/client_builder.py:41]: Hmm, we might need an explicit disconnect here to actually disconnect from the cluster (or we can just wait until the python program exits).


I definitely agree that removing disconnect here is critical because we want to do:
```
client_context = ray.client().connect()
with client_context:
   # do work

with client_context:
   # do some more work
```

File: python/ray/client_builder.py
Comment (ijrsvt) [inline python/ray/client_builder.py:111]:
```suggestion
        if ray.util.client.ray.is_connected() and not self._allow_multiple_cli:
```

**Review by ijrsvt — DISMISSED** [review]

**Comment by ijrsvt** [comment]
I think the overall approach seems fine! Thanks for making this @iycheng :)

**Comment by ericl** [comment]
Hmm a bunch of client unit tests failing, perhaps due to the init hook change.

**Comment by iycheng** [comment]
yes, the init and shutdown issues. I'm fixing this.

File: python/ray/client_builder.py
Comment (iycheng) [reply python/ray/client_builder.py:111]: In this places, it's not entering the multiple cli cases.

We have two type of contexts here: without with statement, it's always the global one.

What it's checking here is that if we previously connected with a single client mode, we are not allowed to connect to multiple cli. So here, it's not `not`

If it's multiple client mode, ray.util.client.ray.is_connect() will always be false. Only cli.is_connect() will return true.

File: python/ray/client_builder.py
Comment (iycheng) [reply python/ray/client_builder.py:112]: see above comment.

File: python/ray/tests/test_client_multi.py
Comment (iycheng) [reply python/ray/tests/test_client_multi.py:27]: It's EOFError right now I think.

File: python/ray/client_builder.py
Comment (ijrsvt) [inline python/ray/client_builder.py:140]: Can we just instantiate a new _ClientContext() here?

**Review by ijrsvt — COMMENTED** [review]
This looks great! I think maybe rebase/merge master to fix some of the test failures?

File: python/ray/client_builder.py
Comment (ericl) [reply python/ray/client_builder.py:116]: I still don't understand, can you provide a code example? Also, the message text should be improved.

Comment (iycheng) [reply python/ray/client_builder.py:116]:
```
ray.init("ray://")

# The following will error
ray.init("ray://", allow_multiple=True)
```

File: python/ray/client_builder.py
Comment (iycheng) [reply python/ray/client_builder.py:140]: The new client context is managed by ray.util.client.ray, so this actually is the new context. It's not reusing anything.

My method here is trying to hide everything from the outside, so all the creation kind of thing will be done in ray.util.client.ray automatically and the outside world just need to set up the context.

If the context is not set up, it'll use the global one (the one created without allow_multiple=True).

File: python/ray/tests/test_client_builder.py
Comment (iycheng) [reply python/ray/tests/test_client_builder.py:237]: Hmmm, somehow I feel I don't need to change it. I'll take another look.

File: python/ray/client_builder.py
Comment (ericl) [reply python/ray/client_builder.py:116]: Thanks! Actually, I think in this case we should allow the connection to proceed. There's no chance of ambiguity here.

File: python/ray/client_builder.py
Comment (ericl) [inline python/ray/client_builder.py:121]:
```suggestion
            raise ValueError("Multiple clients are connected. Additional "
```

**Comment by iycheng** [comment]
The issue seems fixed by this PR (https://github.com/ray-project/ray/issues/17968)
I cherry-picked it and push to test it again. I'll merge after #17968

File: python/ray/client_builder.py
Comment (iycheng) [reply python/ray/client_builder.py]: I won't remove it right now, since it has too many updates. And it also requires API change. I'll try to make them compatible for now:

- single client mode, it will disconnect automatically
- multi client mode, it wont disconnect.

Comment (ericl) [reply python/ray/client_builder.py]: Let me bring this up in api-changes. We definitely don't want two contradicting use cases for with ray.init().

File: python/ray/client_builder.py
Comment (iycheng) [reply python/ray/client_builder.py:116]: there are two things here:
- init
  - with allow_multiple it'll initialize the non-default one
  - without this, it'll initialize the default one
  - if you call `ray.util.client.ray.is_connected` directly without `with` it'll use the default one. so if allow_multiple is set to be true, even you have a handler, it still will return false.
- shutdown
  - if you call ray.shutdown, if it's running without `with`, it'll shutdown all clients. otherwise, it'll just shutdown that context.

I think there are some API not specified clearly: 1) how to define it's connected; 2) which API should apply to all.

**Comment by iycheng** [comment]
I'll wait for CI first and fix other comments.

File: python/ray/client_builder.py
Comment (ericl) [reply python/ray/client_builder.py:116]: Either way, I think we should allow

```
ray.init()
ray.init(allow_multiple=True)
```

The reason is this: say you are writing a library function which uses multiple client connections. Then, your library function cannot be called unless the user of your function passes allow_multiple=True, which is weird.

Make sense?

File: python/ray/tests/test_client_init.py
Comment (ericl) [inline python/ray/tests/test_client_init.py:186]: Why these changes to _disconnect()?

Comment (iycheng) [reply python/ray/tests/test_client_init.py:186]: I previously thought there might be a conflict. Maybe it's not the right fix. I'll follow it up and either give you some explanation or revert it back.

**Comment by iycheng** [comment]
It looks like only Small & Large has some issues. And the wired exception things go away with fixing from @ckw017 

Hopefully, the next update should fix all issues.

File: python/ray/client_builder.py
Comment (iycheng) [reply python/ray/client_builder.py:116]: ok. but how about shutdown and disconnect?

File: python/ray/util/client/__init__.py
Comment (iycheng) [reply python/ray/util/client/__init__.py:204]: Yes, it'll be hit, in multi-threading cases, I think. self._cxt is a thread-local and might not be set up in other threads.  I got errors when run testing.

File: python/ray/tests/test_client_multi.py
Comment (iycheng) [reply python/ray/tests/test_client_multi.py:144]: Nice catch! You remind me of the days when we forgot to add this for all tests in workflow.

Comment (ijrsvt) [reply python/ray/tests/test_client_multi.py:144]: I wish we didn't need to do this--I swear there has been many times this slip has bitten us.

File: python/ray/client_builder.py
Comment (ijrsvt) [reply python/ray/client_builder.py:141]: Ahhh, that makes sense :)

File: python/ray/util/client/__init__.py
Comment (ijrsvt) [reply python/ray/util/client/__init__.py:204]: Hmm okay! What happens if we do multithreading with multiple client contexts? Does that work? Could you maybe test this?

**Comment by iycheng** [comment]
@ericl I'd like to confirm the right behavior of init here:

1. ray.init(, allow_multiple=False)
   - if there is no connection -> proceed (the old case)
   - if there is one connection
       - if it's constructed with allow_multiple=False -> error (it'll error in client api)
       - if it's connected with allow_multiple=True -> proceed.
2. ray.init(, allow_multiple=True)
   - if there is no connection -> proceed
   - if there is connections -> proceed

When we're calling ray.put/ray.get,
- if it's in `with cli`, proceed. 
- if it's not in `with cli`
   - if ray.init(, allow_multiple=False) called, proceed
   - otherwise, error

**Comment by ericl** [comment]
It sounds right, thanks!

File: python/ray/util/client/__init__.py
Comment (iycheng) [reply python/ray/util/client/__init__.py:204]: That should work. I haven't tested it yet. I can add some.

**Comment by iycheng** [comment]
windows test failure looks not related. they have been failed for a while.

File: python/ray/tests/test_client_multi.py
Comment (ericl) [inline python/ray/tests/test_client_multi.py:33]: Hmm in retrospect this is a bit odd. Maybe it should also raise an error (can fix later).

File: python/ray/tests/test_client_multi.py
Comment (ericl) [inline python/ray/tests/test_client_multi.py:58]: Should this raise?

File: python/ray/tests/test_client_multi.py
Comment (ericl) [inline python/ray/tests/test_client_multi.py:61]: Why does this raise but cli2 doesn't?

Comment (ericl) [reply python/ray/tests/test_client_multi.py:61]: Btw, please check the message and type of all exceptions or add a TODO

File: python/ray/tests/test_client_multi.py
Comment (ericl) [reply python/ray/tests/test_client_multi.py:139]: Comment not addressed

File: python/ray/tests/test_client_multi.py
Comment (iycheng) [reply python/ray/tests/test_client_multi.py:58]: @ericl I'm implementing it according to the rules we discussed in the comment.

ray.init(, allow_multiple=False)

=> if there is one connection

=> if it's connected with allow_multiple=True -> proceed.



Could you let me know what's the expected behavior here?

File: python/ray/tests/test_client_multi.py
Comment (iycheng) [reply python/ray/tests/test_client_multi.py:139]: I added comments at L91, do you mean add that for all places?

And L136 is testing get object from another cli. Could you give me more details about what you mean here? Some sample code should be useful.

File: python/ray/tests/test_client_multi.py
Comment (ericl) [reply python/ray/tests/test_client_multi.py:58]: Yep, this was a case I missed. It should raise here.

File: python/ray/tests/test_client_multi.py
Comment (ericl) [reply python/ray/tests/test_client_multi.py:139]: You can use:

```
    with pytest.raises(ValueError, match=r".* Multiple connections blah blah blah .*"):
        myfunc()
```

All the tests in this file can be checking the message (can define the message at top of the unit test file).

Comment (iycheng) [reply python/ray/tests/test_client_multi.py:139]: I think it's not raising these errors right now. It's raising EOFError. I'm not sure how easy right now to detect these errors. I'll just comment all places for now.

**Comment by iycheng** [comment]
@ericl let's revisit the behavior here:

- ray.init(, allow_multiple=False)
  - if there is no connection -> proceed (the old case)
  - if there is one connection
    - if it's constructed with allow_multiple=False -> error (it'll error in client API)
    -**if it's connected with allow_multiple=True -> proceed. ---> error?**

- ray.init(, allow_multiple=True)
  - if there is no connection -> proceed
  - if there is connections
    -   **if it's inited with ray.init(, allow_multiple=False) ---> what to do ?**
    -  otherwise -> proceed

When we're calling ray.put/ray.get,
- if it's in with cli, proceed.
- if it's not in with cli
  - if ray.init(, allow_multiple=False) called, proceed
   - otherwise, error

**Comment by ericl** [comment]
Only change should be the one I requested in comments

**Comment by iycheng** [comment]
I think it's a little bit wired then, 
In your comment you mean this one should raise error:
```
ray.init(, allow_multiple=False)
ray.init(, allow_multiple=True)
```
will raise errors. But if I do nothing, this will not throw error:

```
ray.init(, allow_multiple=True)
ray.init(, allow_multiple=False)
```
This is a little bit wired, right? SHould the later case also raise an error?

**Comment by ericl** [comment]
I believe the comment is saying the latter only should raise an error.

File: python/ray/tests/test_client_multi.py
Comment (ijrsvt) [inline python/ray/tests/test_client_multi.py:180]: I think a failure in the thread won't actually fail the test. Can you maybe have the thread push to a shared queue on success and assert on the main thread that two entries are in the queue?

File: python/ray/tests/test_client_multi.py
Comment (ijrsvt) [inline python/ray/tests/test_client_multi.py]: Awesome! Thanks!!!

**Review by ericl — APPROVED** [review]

**Pull request merged**
